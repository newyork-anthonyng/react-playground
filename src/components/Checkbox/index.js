import React, { Component } from 'react';
import T from 'prop-types';
import noop from '../../utility/noop';

class Checkbox extends Component {
  handleKeyPress = (e) => {
    // check for ENTER (13) or SPACEBAR (32)
    const isSelected = /^(13|32)/.test(e.charCode);

    if (isSelected) {
      this.props.onChange();
    }
  }

  handleClick = () => {
    this.props.onChange();
  }

  render() {
    const { checked, label } = this.props;

    return (
      <div
        role="checkbox"
        tabIndex="0"
        aria-checked={`${checked}`}
        aria-label={label}
        onKeyPress={this.handleKeyPress}
        onClick={this.handleClick}
        style={{
          border: '1px solid black',
          display: 'inline-block',
          height: '25px',
          width: '25px',
          position: 'relative',
          cursor: 'pointer',
          userSelect: 'none',
          borderRadius: '6px',
        }}
      >
        <span style={{
          position: 'absolute',
          top: '3px',
          left: '6px',
        }}>{checked && 'âœ“'}</span>
      </div>
    );
  }
}

Checkbox.defaultProps = {
  checked: false,
  onChange: noop,
  label: 'Unlabelled checkbox',
};

Checkbox.propTypes = {
  checked: T.bool.isRequired,
  onChange: T.func.isRequired,
  label: T.string.isRequired,
};

export default Checkbox;
