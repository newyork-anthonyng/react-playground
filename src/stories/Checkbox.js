import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, number, boolean, text } from '@storybook/addon-knobs';

import Checkbox from '../components/Checkbox';

storiesOf('Checkbox', module)
  .addDecorator(withKnobs)
  .add('With knobs', () => {
    return (
      <Checkbox
        checked={boolean('Checked', false)}
        onChange={action('Checkbox changed')}
        label={text('Checkbox label', 'My awesome checkbox')}
      />
    );
  })
  .add('Example', () => {
    class CheckboxContainer extends React.Component {
      constructor() {
        super();

        this.state = {
          checked: false,
          label: '',
        };
      }

      handleTextChange = (e) => {
        this.setState({ label: e.target.value });
      }

      handleCheckboxChange = () => {
        this.setState({
          checked: !this.state.checked,
        });
      }

      render() {
        return (
          <div>
            <input
              type="text"
              value={this.state.label}
              onChange={this.handleTextChange}
              placeholder="Type label for the checkbox..."
              style={{ display: 'block' }}
            />
            <Checkbox
              label={this.state.label}
              onChange={this.handleCheckboxChange}
              checked={this.state.checked}
            />
          </div>
        );
      }
    }

    return (
      <CheckboxContainer />
    );
  });
