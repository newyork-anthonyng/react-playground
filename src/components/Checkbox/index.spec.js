import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import noop from "../../utility/noop";
import Checkbox from "./";

const defaultProps = {
  checked: false,
  onChange: noop,
  label: "Unlabelled checkbox"
};

it("should render correctly when checked", () => {
  const wrapper = shallow(<Checkbox {...defaultProps} checked />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it("should render correctly when unchecked", () => {
  const wrapper = shallow(<Checkbox {...defaultProps} checked={false} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it("should run callback when ENTER is pressed", () => {
  const cb = jest.fn();
  const wrapper = shallow(<Checkbox {...defaultProps} onChange={cb} />);

  wrapper.simulate("keypress", { charCode: 13 });

  expect(cb).toHaveBeenCalledTimes(1);
});

it("should run callback when SPACEBAR is pressed", () => {
  const cb = jest.fn();
  const wrapper = shallow(<Checkbox {...defaultProps} onChange={cb} />);

  wrapper.simulate("keypress", { charCode: 32 });

  expect(cb).toHaveBeenCalledTimes(1);
});

it("should run callback when clicked", () => {
  const cb = jest.fn();
  const wrapper = shallow(<Checkbox {...defaultProps} onChange={cb} />);

  wrapper.simulate("click");

  expect(cb).toHaveBeenCalledTimes(1);
});
