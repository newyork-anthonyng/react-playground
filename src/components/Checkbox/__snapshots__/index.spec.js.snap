// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly when checked 1`] = `
<div
  aria-checked="true"
  aria-label="Unlabelled checkbox"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="checkbox"
  style={
    Object {
      "border": "1px solid black",
      "borderRadius": "6px",
      "cursor": "pointer",
      "display": "inline-block",
      "height": "25px",
      "position": "relative",
      "userSelect": "none",
      "width": "25px",
    }
  }
  tabIndex="0"
>
  <span
    style={
      Object {
        "left": "6px",
        "position": "absolute",
        "top": "3px",
      }
    }
  >
    âœ“
  </span>
</div>
`;

exports[`should render correctly when unchecked 1`] = `
<div
  aria-checked="false"
  aria-label="Unlabelled checkbox"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="checkbox"
  style={
    Object {
      "border": "1px solid black",
      "borderRadius": "6px",
      "cursor": "pointer",
      "display": "inline-block",
      "height": "25px",
      "position": "relative",
      "userSelect": "none",
      "width": "25px",
    }
  }
  tabIndex="0"
>
  <span
    style={
      Object {
        "left": "6px",
        "position": "absolute",
        "top": "3px",
      }
    }
  />
</div>
`;
